packages/controllers/src/cronjob/CronjobController.ts

```diff
@@ -0,0 +1,352 @@
+      'SnapController:snapUpdated',
+      this._handleEventSnapUpdated,
+    );
+  }
+
+  /**
+   * Retrieve all cronjob specifications for all runnable snaps.
+   *
+   * @returns Array of Cronjob specifications.
+   */
+  private async getAllJobs(): Promise<Cronjob[]> {
+    const snaps = await this.messagingSystem.call('SnapController:getAll');
+    const filteredSnaps = getRunnableSnaps(snaps);
+
+    const jobs = await Promise.all(
+      filteredSnaps.map((snap) => this.getSnapJobs(snap.id)),
+    );
+    return flatten(jobs).filter((job) => job !== undefined) as Cronjob[];
+  }
+
+  /**
+   * Retrieve all Cronjob specifications for a Snap.
+   *
+   * @param snapId - ID of a Snap.
+   * @returns Array of Cronjob specifications.
+   */
+  private async getSnapJobs(snapId: SnapId): Promise<Cronjob[] | undefined> {
+    const permissions = await this.#messenger.call(
+      'PermissionController:getPermissions',
+      snapId,
+    );
+
+    const permission = permissions?.[SnapEndowments.Cronjob];
+    const definitions = getCronjobCaveatJobs(permission);
+
+    return definitions?.map((definition, idx) => {
+      return { ...definition, id: `${snapId}-${idx}`, snapId };
+    });
+  }
+
+  /**
+   * Register cron jobs for a given snap by getting specification from a permission caveats.
+   * Once registered, each job will be scheduled.
+   *
+   * @param snapId - ID of a snap.
+   */
+  async register(snapId: SnapId) {
+    const jobs = await this.getSnapJobs(snapId);
+    jobs?.forEach((job) => this.schedule(job));
+  }
+
+  /**
+   * Schedule a new job.
+   * This will interpret the cron expression and tell the timer to execute the job
+   * at the next suitable point in time.
+   * Job last run state will be initialized afterwards.
+   *
+   * Note: Schedule will be skipped if the job's execution time is too far in the future and
+   * will be revisited on a daily check.
+   *
+   * @param job - Cronjob specification.
+   */
+  private schedule(job: Cronjob) {
+    if (this.#timers.has(job.id)) {
+      return;
+    }
+    const parsed = parseCronExpression(job.expression);
+    const next = parsed.next();
+    const now = new Date();
+    const ms = next.getTime() - now.getTime();
+
+    // Don't schedule this job yet as it is too far in the future
+    if (ms > DAILY_TIMEOUT) {
+      return;
+    }
+
+    const timer = new Timer(ms);
+    timer.start(() => {
+      this.executeCronjob(job);
+      this.#timers.delete(job.id);
+      this.schedule(job);
+    });
+
+    this.updateJobLastRunState(job.id, 0); // 0 for init, never ran actually
+    this.#timers.set(job.id, timer);
+    this.#snapIds.set(job.id, job.snapId);
+  }
+
+  /**
+   * Execute job.
+   *
+   * @param job - Cronjob specification.
+   */
+  private executeCronjob(job: Cronjob) {
+    this.updateJobLastRunState(job.id, Date.now());
+    this.#messenger.call('SnapController:handleRequest', {
+      snapId: job.snapId,
+      origin: '',
```

Comment:
Can we remove this? Or should it maybe be something like `'cron'`?

Reply:
We currently use empty string for transaction insights as well - It is currently required to have a value by the snap executor. We should perhaps figure out a way to only require it for certain handlers? Out of scope for this PR imo.